### ISSUE TITLE
Question about the computation of cdf in the code

### ISSUE THREAD

**Issue by yaoqi-zd** [fab-jul/L3C-PyTorch#18]
Hi, I'm curious about the computation of cdf in your code. The quantization levels have length of L(default 25 in your code), why the cdf takes the length of (L+1) for each quantized value?

for uniform distribution, you seem to pad zeros to the new added dimension, like:
```
# here cdf's shape is [N, H, W, L]
cdf = torch.cumsum(pr, -1)
cdf = cdf.mul_(2**precision)
cdf = cdf.round()
# here cdf's shape changes to [N, H, W, L+1]
cdf = torch.cat((torch.zeros((N, H, W, 1), dtype=cdf.dtype, device=cdf.device),
                     cdf), dim=-1)
```
and for mixture of logistic, you seem to re-define new quantization levels, like:
```
# Lp = L+1
# here new quantization levels are defined
self.targets = torch.linspace(dmll.x_min - dmll.bin_width / 2,
                                      dmll.x_max + dmll.bin_width / 2,
                                      dmll.L + 1, dtype=torch.float32, device=l.device)
```

**Comment by fab-jul** [comment]
Hi, we use the CDF to calculate the PMF. For example, let's say `x` takes integer values. We would then calculate `p(x) = cdf(x+1/2) + cdf(x-1/2)`. Note that if `x` takes `L` values, we would need to evaluate the CDF at `L+1` values.

For the latents, `x` takes one of `L` values in `[-1,1]`, so values are `bin_width = 2/(L-1)` apart, and we thus use `bin_width/2` to get `p(x) = cdf(x+bin_width/2) + cdf(x-bin_width/2)`. 

Does that help?

**Comment by yaoqi-zd** [comment]
@fab-jul Thanks for your reply!
1) so the function `any_backend.encode_cdf` can take any valid cdf + sym as input and output the byte stream?
2) if I want to encode discrete variable `X` which takes values `[-1, 0, 1]`, I need to compute cdf function values `cdf` at `targets=[-1.5, -0.5, 0.5, 1.5]`, then send `cdf` and `sym` into function `any_backend.encode_cdf(cdf, sym)`. Am I right?

**Comment by yaoqi-zd** [comment]
@fab-jul Hi, I find not all cdf can result in correct decoded results, a simple test snippet likes:
```
    import numpy as np
    import torch
    from torchac import torchac

    # create cdf variable
    cdf = np.array([0, 0, 1., 1.]).reshape((1, 1, 1, 4))
    cdf_t = torch.from_numpy(cdf)
    cdf_t = cdf_t.repeat((1, 2, 2, 1))  # repeat to shape: [1, 2, 2, 4]
    
    # create symbols, where prob(sym=1)=1
    sym = np.array([[1, 1], [1, 1]]).reshape((1, 2, 2))
    sym_t = torch.from_numpy(sym).to(torch.int16)  
    
    # cast cdf to int16
    precision = 16
    cdf_t = cdf_t.mul_(2**precision)
    cdf_t = cdf_t.round()
    cdf_t = cdf_t.to(torch.int16)
    
    # encode and write bytes
    enc_bytes = torchac.encode_cdf(cdf_t, sym_t.flatten())
    num_bytes = np.array(len(enc_bytes), dtype=np.uint32)

    fp = open("enc_dec.bin", "wb")
    fp.write(num_bytes.tobytes())
    fp.write(enc_bytes)
    fp.close()
   
    # read bytes and decode
    fp = open("enc_dec.bin", "rb")
    num_bytes = np.frombuffer(fp.read(4), dtype=np.uint32)[0]
    encoded = fp.read(num_bytes)
    decoded = torchac.decode_cdf(cdf_t, encoded)
    decoded = decoded.reshape(1, 2, 2)
    print(decoded)
```
the printed `decoded` is [[2, 2], [2,2]], so how can we handle such situation when the pdf is concentrated on a single symbol?

what's more, I see you renormalize the cdf in torchac.py
```
def _renorm_cast_cdf_(cdf, precision):
    Lp = cdf.shape[-1]
    finals = 1  # NHW1
    # RENORMALIZATION_FACTOR in cuda
    f = torch.tensor(2, dtype=torch.float32, device=cdf.device).pow_(precision)
    cdf = cdf.mul((f - (Lp - 1)) / finals)  # TODO
    cdf = cdf.round()
    cdf = cdf.to(dtype=torch.int16, non_blocking=True)
    r = torch.arange(Lp, dtype=torch.int16, device=cdf.device)
    cdf.add_(r)
    return cdf
```
what's the purpose of multiplying `f-(Lp-1) / finals` rather than `f`, and `cdf.add_(r)`?

Thanks
Qi Yao

**Comment by fab-jul** [comment]
So, there are 2 problems, somewhat unrelated to your example:

1. the problem is that I didn't write the arithmetic coding to handle symbols with 0 probability. After all, even if you predict a symbol to have 0 probability, it may still show up in your datastream, and then you would need infinite bits to encode it. Thus, CDFs that are not strictly monotonically increasing, like [0, 0, 1, 1], are not supported.
2. arithmetic coding works in integer space, and uses int16 to encode the CDF.

Both of these problems are why we have `_renorm_cast_cdf_`. The high level idea is to add the minimal amount of pertubation to the real CDF to get a CDF that is strictly monotonically increasing. This minimal amount is defined by the arithmetic coding implementation, which uses uint16 to encode the CDF. So, first of all, we look at the CDF not in the range [0, 1], but in the integer range `{0, ..., 2**16}`. So, the minimal amount you can add to one element of the CDF is `1/(2**16)` in float space. In integer space, you just add 1 to the first symbol, 2 to the second, and so on. If you have `Lp` symbols, you add `Lp` to the CDF of the last symbol. However, now you are not in range `{0, ..., 2**16}` anymore.

To fix all this, we have `_renorm_cast_cdf_`. It first multiplies the float CDF with `2**16 - (Lp - 1)`. Now we have values in `{0, ..., 2**16 - (Lp - 1)}`, which in general are *not* monotonically increasing. That's why we add `arange(Lp)`, which is masically a line with slope 1. And voila, we have a strictly monotonically increasing CDF in integer space, with values in `{0, ..., 2**16 - (Lp - 1)}`.

---

With all that said, I think your code still *could* or maybe even should work, but I don't remember enough implementation details to say for sure why it doesn't. Note that e.g. `cdf = np.array([0, 0.001, 0.009, 1.])` works.

**Comment by fab-jul** [comment]
Ahh I saw the problem now! It's in the python code above and due to PyTorch not having uint16 (I think?):

```
    # cast cdf to int16
    precision = 16
    cdf_t = cdf_t.mul_(2**precision)
    cdf_t = cdf_t.round()  # [0., 0., 65536., 65536.]
    cdf_t = cdf_t.to(torch.int16)  # [0., 0., 0., 0.]
``` 

SO, your 1s get mapped to 0s.

I remember that I sadly had to using int16 because PyTorch doesn't have uint16.

**Comment by yaoqi-zd** [comment]
@fab-jul Thank you so much for your detailed explanations. Finally I can use your `any_backend.encode_cdf` function to compress my data!

**Comment by fab-jul** [comment]
sounds good! let me know if you find any other problems. also, do you think I should release torchac as a standalone package?

**Comment by yaoqi-zd** [comment]
@fab-jul actually I hope you could release torchac as a standalone package, since [ range coder](https://tensorflow.github.io/compression/docs/range_coding.html) package used in tensorflow is not easy to re-implement. `torchac` is the most convenient library I've ever found and can reduce the effort for those new to the data compression field.

**Comment by fab-jul** [comment]
cool! yeah I have the same feeling but didn’t invest the time yet. would you be available for some testing?

**Comment by yaoqi-zd** [comment]
sadly I have to work on weekday, and if you don't mind, I can spend my weekend time to help the testing

**Comment by fab-jul** [comment]
yeah same - I'll see when I have some time and post a link here.

**Comment by yaoqi-zd** [comment]
ok

**Comment by trougnouf** [comment]
Hello @fab-jul ,

I have been struggling to understand how to use your range_encoder/range_decoder (`encode_cdf`/`decode_cdf`) functions.

I am coming from liujiaheng's code, where quantization is equivalent to rounding and the cdf is a function; either a distribution conditioned on a per patch sigma, or a learned prior ("`bitEstimator`", which ends in a sigmoid, https://github.com/liujiaheng/compression/blob/master/models/bitEstimator.py ). This yields something like
```
prob = bitEstimator(z + 0.5) - bitEstimator(z - 0.5)
total_bits = torch.sum(torch.clamp(-1.0 * torch.log(prob + 1e-10) / math.log(2.0), 0, 50))
```

`z` has dimensions `BS,C,H,W`, the dimensions can be any except for the number of channels `C` which must match because stacks of channels are being encoded and the prior has per channel parameters.

If the cdf is to be encoded in a table, which seems to be required for `encode_cdf`'s input, then I have to encode every possible value (`range(min_value, max_value, precision)`, which I guess is the `L` dimension) for every prior (`BS, H, W`) for every channel (`C`). 

I can do something like that (assuming that encoded values must be positive):
```
ntargets = int((-min_feat+max_feat)/q_intv+1)
targets = torch.linspace(min_feat, max_feat, ntargets)
nchan = z.size(1)
targets = targets.expand(nchan, ntargets).unsqueeze(0).unsqueeze(-1)
cdf_tbl = cdf_fun(targets).reshape(nchan,1,1,ntargets)
# or cdf_tbl = cdf_fun(targets).squeeze(-1).unsqueeze(-2) for [1,C,1,L]
cdf_tbl = (cdf_tbl*(2**precision)).short()
z_int = ((z-min_feat)/q_intv).round().short()
```
to get a` [nchan,1,1,ntargets]` integer `cdf_tbl` and a `[bs,nchan,H,W]` `z_int` (features)

where I am confused is:

**The required format is `N,H,W,Lp`, I think `N` is the channel dimension, but the function `encode_cdf(cdf, sym)` expects a 1HWLp cdf, so presumably a single channel at a time.** Is it supposed to encode one channel at a time for the whole image? Does that work when the priors may be different for different features?

I tried using another dimension as the channel and encoding a feature, but I get wrong results

```
z_int = ((z-min_feat)/q_intv).round().short()#*(2**precision)).short()
print(z_int.min())
>>>tensor(75, dtype=torch.int16)
a_feature = z_int[0,:,2,2]
print(a_feature)
>>>tensor([126, 126, 126, 126, 126, ...], dtype=torch.int16)
encoded_str = torchac.encode_cdf(cdf_tbl, a_feature)
print(encoded_str)
>>>b'?\xdc@\x1f\xff\x85S@\x00\x07\xff\xc0\x07\xc5!\xf9\xc0\x00\x00\x00\x00\xd6b\x9al\x00\x0f\x9e\x88\x00\x11\xc1\x0e\x18\x1f\x99\xc3\xff\xf0\x00\x00\xef\x82\x01\xe5_V\xaa@\x1f\xff\x80<nz\xb6\x02!\x80\x00\x00?\xfd\x80\xe6|\x004\xbbZ{\x01\xf9V\xd8\xa0"h\x00\xaf\x0c{<\x00\t6\x80\x0b\x17\x00\x0b\n\x00\x00H\xf1\x01\xd9\x0c@\x00\xdfH?\xa6\x84\x00A\xff\xf8\x08'
decoded_str = torchac.decode_cdf(cdf_tbl, encoded_str)
print(decoded_str)
>>>tensor([126, 126, 125, 253, 253,...], dtype=torch.int16)
```
